BlueprintUICommonControls/Sources/AttributedLabel.swift

```diff
@@ -43,53 +57,508 @@ public struct AttributedLabel: Element, Hashable {
+        /// `accessibilityCustomRotors` getter.
+        private var accessibilityLinks: [LinkAccessibilityElement] = []
+
+        override var accessibilityCustomRotors: [UIAccessibilityCustomRotor]? {
+            set { fatalError() }
+            get {
+                [
+                    UIAccessibilityCustomRotor(systemType: .link) { [weak self] predicate in
+                        guard let self = self else {
+                            return nil
+                        }
+
+                        self.accessibilityLinkIndex += predicate.searchDirection == .next ? 1 : -1
+                        self.accessibilityLinkIndex = min(self.accessibilityLinks.count - 1, self.accessibilityLinkIndex)
+                        self.accessibilityLinkIndex = max(0, self.accessibilityLinkIndex)
+
+                        let link = self.accessibilityLinks[self.accessibilityLinkIndex]
+                        return UIAccessibilityCustomRotorItemResult(targetElement: link, targetRange: nil)
+                    },
+                ]
+            }
+        }
+
+        var linkHandler: LinkHandler?
+
+        func update(model: AttributedLabel, linkHandler: LinkHandler?, isMeasuring: Bool) {
+            let previousString = attributedText?.string
+
+            linkAttributes = model.linkAttributes
+            activeLinkAttributes = model.activeLinkAttributes
+            linkDetectionTypes = model.linkDetectionTypes ?? []
+            attributedText = model.attributedText
+                .applyingDefaultFont()
+                .replacingLinkAttributes()
+            numberOfLines = model.numberOfLines
+            textRectOffset = model.textRectOffset
+            isAccessibilityElement = model.isAccessibilityElement
+            self.linkHandler = linkHandler
+
+            if !isMeasuring, previousString != attributedText?.string {
+                links = attributedLinks(in: model.attributedText) + detectedDataLinks(in: model.attributedText)
+                accessibilityLinks = accessibilityLinks(for: links, in: model.attributedText)
+            }
+
+            applyLinkColors()
+        }
+
+        private func updateAccessibilityTraits(_ model: AttributedLabel) {
+            if let traits = model.accessibilityTraits {
+                var union = accessibilityTraits.union(UIAccessibilityTraits(with: traits))
+                // UILabel has the `.staticText` trait by default. If we explicitly set `.updatesFrequently` this should be removed.
+                if traits.contains(.updatesFrequently) && accessibilityTraits.contains(.staticText) {
+                    union.subtract(.staticText)
+                }
+                accessibilityTraits = union
+            }
+        }
+
         override func drawText(in rect: CGRect) {
             super.drawText(in: rect.offsetBy(dx: textRectOffset.horizontal, dy: textRectOffset.vertical))
         }
+
+        func makeTextStorage() -> NSTextStorage? {
```

Comment:
Do we only make these if there's links to render?

Reply:
Yes